morango/sync/syncsession.py

```diff
@@ -376,42 +384,139 @@ def _pull_record_chunk(self, chunk_size, transfer_session):
+    """The in progress signal, which should be fired at least once during the procedure."""
+    completed = SyncSignal()
+    """The completed signal, which should be fired at the end of the procedure"""
+
+    def __init__(self, **kwargs_defaults):
+        super(SyncSignalGroup, self).__init__(**kwargs_defaults)
+        self.started = SyncSignal(**kwargs_defaults)
+        self.in_progress = SyncSignal(**kwargs_defaults)
+        self.completed = SyncSignal(**kwargs_defaults)
+
+        self.started.connect(self.fire)
+        self.in_progress.connect(self.fire)
+        self.completed.connect(self.fire)
+
+    def send(self, **kwargs):
+        """
+        Context manager helper that will signal started and fired when entered and exited,
+        and it'll fire those with the `kwargs`.
+
+        :rtype: SyncSignalGroup
+        """
+        defaults = self._defaults.copy()
+        defaults.update(kwargs)
+        context_group = SyncSignalGroup(**defaults)
+        context_group.started.connect(self.started.fire)
+        context_group.in_progress.connect(self.in_progress.fire)
+        context_group.completed.connect(self.completed.fire)
+        return context_group
+
+    def __enter__(self):
+        """
+        Fires the `started` signal.
+        """
+        self.started.fire()
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        """
+        Fires the `completed` signal.
+        """
+        self.completed.fire()
+
+
 class SyncClient(object):
     """
     Controller to support client in initiating syncing and performing related operations.
     """
 
+    session = SyncSignalGroup()
+    """Signal group firing for each push and pull `TranferSession`."""
+    queuing = SyncSignalGroup(local=True)
+    """Queuing signal group for locally or remotely queuing data before transfer."""
+    pushing = SyncSignalGroup(transfer_session=None)
+    """Pushing signal group for tracking push progress of `TransferSession`. It is sent the `TransferSession` model."""
+    pulling = SyncSignalGroup(transfer_session=None)
+    """Pulling signal group for tracking pull progress of `TransferSession`. It is sent the `TransferSession` model."""
+    dequeuing = SyncSignalGroup(local=True)
+    """Dequeuing signal group for locally or remotely dequeuing data after transfer."""
+
     def __init__(self, sync_connection, sync_session, chunk_size=500):
+        """
+        :type sync_connection: NetworkSyncConnection
+        :type sync_session: SessionWrapper
+        :type chunk_size: int
+        """
         self.sync_connection = sync_connection
         self.sync_session = sync_session
         self.chunk_size = chunk_size
         self.current_transfer_session = None
 
     def initiate_push(self, sync_filter):
-
-        logger.info("Initiating push sync")
         self._create_transfer_session(True, sync_filter)
 
-        _queue_into_buffer(self.current_transfer_session)
+        with self.queuing.send(local=True):
```

Comment:
I did consider passing the transfer session to all signals, since we could infer local/remote based off that object. It wasn't required for the implementation so I avoided adding it. 

I'm not sure I understand how multiple syncs would occur for the client. If you're talking about a push AND pull for a client, then I can see that but I still don't see how that would happen in parallel.

Reply:
It probably won't be possible via Kolibri's UI, given how we process the job queue serially, but we can also initiate this syncing via management commands, and could initiate many in parallel. But since each of those is starting up in a separate process, I assume these signals would be contextualized to the appropriate task.